Sources/PrimerSDK/Classes/User Interface/Apple Pay/ApplePayViewModel.swift

```diff
@@ -1,69 +1,317 @@
+                                                            Primer.shared.presentingViewController?.present(Primer.shared.root!, animated: true)
+                                                        }
+                                                    })
+                                                    
+                                                    completion(err)
+                                                }
+                                            })
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    
+                case .failure(let err):
+                    completion(err)
+                }
+            }
+            // FIXME: Present on a window above the app's window
+            (Primer.shared.delegate as? UIViewController)?.present(paymentVC, animated: true, completion: nil)
+            
+        } else {
+            log(logLevel: .error, title: "APPLE PAY", message: "Cannot make payments on the provided networks")
+            return completion(AppleException.unableToMakePaymentsOnProvidedNetworks)
+        }
+    }
+    // swiftlint:enable cyclomatic_complexity function_body_length
+}
+
+extension ApplePayViewModel: PKPaymentAuthorizationViewControllerDelegate {
+    
+    func paymentAuthorizationViewControllerDidFinish(_ controller: PKPaymentAuthorizationViewController) {
+        applePayCompletion?(.failure(AppleException.cancelled))
+        applePayCompletion = nil
+    }
+    
+    @available(iOS 11.0, *)
+    func paymentAuthorizationViewController(
+        _ controller: PKPaymentAuthorizationViewController,
+        didAuthorizePayment payment: PKPayment,
+        handler completion: @escaping (PKPaymentAuthorizationResult) -> Void
+    ) {
+        do {
+            let tokenPaymentData = try JSONParser().parse(ApplePayPaymentResponseTokenPaymentData.self, from: payment.token.paymentData)
+            let applePayPaymentResponse = ApplePayPaymentResponse(
+                token: ApplePayPaymentResponseToken(
+                    paymentMethod: ApplePayPaymentResponsePaymentMethod(
+                        displayName: payment.token.paymentMethod.displayName,
+                        network: payment.token.paymentMethod.network?.rawValue,
+                        type: payment.token.paymentMethod.type.primerValue
+                    ),
+                    transactionIdentifier: payment.token.transactionIdentifier,
+                    paymentData: tokenPaymentData
+                )
+            )
+
+            applePayCompletion?(.success(applePayPaymentResponse))
+            applePayCompletion = nil
+        } catch {
+            applePayCompletion?(.failure(error))
+            applePayCompletion = nil
+        }
+    }
+    
+}
+
+extension PKPaymentMethodType {
```

Comment:
[Q]
Is it a good iOS practice to extend these types?

Reply:
Well, it's very common when you want to add functionality on existing classes. This is not the same case as with the conflicts that Voi encountered, since `PKPaymentMethodType` cannot be subclassed. I just modified the access level to internal.

Also, since I've chosen a name that starts with `primer`, I consider it quite difficult another developer to want to have the same extension name.